contracts/vesting/src/contract.rs

```diff
@@ -0,0 +1,636 @@
+        from: Address,
+        to: Address,
+        amount: i128,
+    ) -> Result<(), ContractError> {
+        from.require_auth();
+
+        if amount <= 0 {
+            log!(&env, "Vesting: Transfer token: Invalid transfer amount");
+            panic_with_error!(env, ContractError::InvalidTransferAmount);
+        }
+
+        let token_client = token_contract::Client::new(&env, &get_token_info(&env).address);
+
+        verify_vesting(&env, &from, amount, &token_client)?;
+        token_client.transfer(&from, &to, &amount);
+
+        env.events().publish(
+            (
+                "Transfer token",
+                "Transfering tokens between accounts: from: {}, to:{}, amount: {}",
+            ),
+            (from, to, amount),
+        );
+
+        Ok(())
+    }
+
+    fn transfer_vesting(
+        env: Env,
+        from: Address,
+        to: Address,
+        amount: i128,
+        curve: Curve,
+    ) -> Result<(), ContractError> {
+        from.require_auth();
+
+        let whitelist = get_whitelist(&env);
+        if !whitelist.contains(from.clone()) {
+            log!(
+                &env,
+                "Vesting: Transfer Vesting: Not authorized to transfer vesting"
+            );
+            panic_with_error!(env, ContractError::NotAuthorized);
+        }
+
+        if amount <= 0 {
+            log!(
+                &env,
+                "Vesting: Transfer Vesting: Transfer amount must be positive"
+            );
+            panic_with_error!(env, ContractError::InvalidTransferAmount);
+        }
+
+        assert_schedule_vests_amount(&env, &curve, amount)?;
+
+        // if not fully vested we update
+        if curve.value(env.ledger().timestamp()) != 0 {
+            update_vesting(&env, &to, amount, curve)?;
+        }
+
+        let token_client = token_contract::Client::new(&env, &get_token_info(&env).address);
+
+        verify_vesting(&env, &from, amount, &token_client)?;
+        token_client.transfer(&from, &to, &amount);
+
+        env.events().publish(
+            (
+                "Transfer vesting",
+                "Transfering vesting between accounts: from: {}, to: {}, amount: {}",
+            ),
+            (from, to, amount),
+        );
+
+        Ok(())
+    }
+
+    fn burn(env: Env, sender: Address, amount: i128) -> Result<(), ContractError> {
+        sender.require_auth();
+
+        if amount <= 0 {
+            log!(&env, "Vesting: Burn: Invalid burn amount");
+            panic_with_error!(env, ContractError::InvalidBurnAmount);
+        }
+
+        let remainder = get_vesting_total_supply(&env) - amount;
+        match remainder < 0 {
+            true => {
+                log!(
+                    &env,
+                    "Vesting: Burn: Critical error - total supply cannot be negative"
+                );
+                panic_with_error!(env, ContractError::Std);
+            }
+            false => update_vesting_total_supply(&env, remainder),
+        };
```

Comment:
I don't know why I didn't pick up that sooner, this one is a pretty obvious one.
```suggestion
        if remainder < 0 {
            log!(
                &env,
                "Vesting: Burn: Critical error - total supply cannot be negative"
            );
            panic_with_error!(env, ContractError::Std);
        } else {
            update_vesting_total_supply(&env, remainder);
        };
```
or even better
```suggestion
        if remainder >= 0 {
            update_vesting_total_supply(&env, remainder);
        } else {
            log!(
                &env,
                "Vesting: Burn: Critical error - total supply cannot be negative"
            );
            panic_with_error!(env, ContractError::Std);
        };
```